samcli/lib/pipeline/bootstrap/resource.py

```diff
@@ -0,0 +1,133 @@
+""" Represents AWS resource"""
+from typing import List, Optional
+
+
+class ARNParts:
+    """
+    Decompose a given ARN into its parts
+
+    Attributes
+    ----------
+    partition: str
```

Comment:
I followed [this](https://realpython.com/documenting-python-code/) format assuming it is standard.

Reply:
You can see some examples here without types: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

In any case this is error-prone, as shown by e.g. `ARNParts` or `Stage` where the types are incorrect.

Reply:
Where is the incorrect types in `ARNParts` and `Stage`, I didn't find it?

I'm trying to see what is the standards; I see in our code base we set the parameter's type in the docs. Also, even without type, it is still error-prone if you added/removed/renamed a parameter without modifying the docs accordingly. So the question is whether fixing the discrepancy is the correct way or not.

Reply:
In `ARNParts` the names are wrong (e.g. `account-id` instead of `account_id`), and in `Stage` the types are wrong (e.g. `aws_profile: str` in docstring and `aws_profile: Optional[str]` in type hint).

Ensuring type hints and docstring have same arguments is one thing; ensuring types are correct is yet another thing on top of that.

Reply:
Why do you think they are different? it is the same exact issue; discrepancies between the docs and the actual arguments.

Reply:
Sure, they're both a discrepancy. But duplicated types requires ensuring (1) the names match, and (2) the types match, while not specifying type in docstring only requires (1).

Reply:
@hoffa and I had this chat before about whether or not to include types in docstring. The conclusion was just to not "have" to because it will increase the maintenance cost and it does not bring any values in modern IDEs (IDE will pick up types from signatures).

Reply:
then while reading the docs (running `help(X)`) of a class or method with many attributes, we will need to keep jumping over to the top of the docs to know the type of an argument that is documented all the way down in the list.

Reply:
Generally we would avoid to have too many attributes (7 is the limit by pylint), but that's a fair point.

Reply:
In any case the types in the `Stage` docstring are wrong, which is my point. Too easy to make a mistake.

Reply:
let me fix the wrong types; Removing all of the types should be a team decision and applied in a separate command as it is:
1. Will affect everywhere; most of the docstrong across the package include type-hint.
2. Seems to be not the standard way.
2.